// @questlabs/auth-js/dist/index.mjs
// src/core.ts
var EXPECTED_MESSAGE_TYPE = "authorization_response";
var DEFAULT_OAUTH_BROKER_URL = "https://oauth.greta.sh/~oauth/initiate";
var DEFAULT_SUPPORTED_OAUTH_ORIGINS = ["https://oauth.greta.sh"];
var DEFAULT_MOBILE_DEEP_LINK_REDIRECT_URI = "greta://oauth-callback";
var DEFAULT_DESKTOP_LOCALHOST_REDIRECT_URI = "http://127.0.0.1/iframe-oauth/callback";
var POPUP_CHECK_INTERVAL_MS = 500;
var IFRAME_FALLBACK_TIMEOUT_MS = 12e4;
function generateState() {
  if (typeof crypto !== "undefined" && crypto.getRandomValues) {
    return [...crypto.getRandomValues(new Uint8Array(16))].map((b) => b.toString(16).padStart(2, "0")).join("");
  }
  return Math.random().toString(36).substring(2) + Date.now().toString(36);
}
function isDevice() {
  const ua = navigator.userAgent;
  if (/iPhone|iPad|iPod|Android/i.test(ua)) return true;
  if (/Macintosh/i.test(ua) && navigator.maxTouchPoints > 1) return true;
  return false;
}
function getPopupDimensions(inIframe) {
  const hasBrowserPosition = window.screenX !== 0 || window.screenY !== 0 || !inIframe;
  const width = hasBrowserPosition ? window.outerWidth * 0.5 : window.screen.width * 0.5;
  const height = hasBrowserPosition ? window.outerHeight * 0.5 : window.screen.height * 0.5;
  const left = hasBrowserPosition ? window.screenX + (window.outerWidth - width) / 2 : (window.screen.width - width) / 2;
  const top = hasBrowserPosition ? window.screenY + (window.outerHeight - height) / 2 : (window.screen.height - height) / 2;
  return { width, height, left, top };
}
function startWebMessageListener(supportedOrigins) {
  let resolvePromise;
  const promise = new Promise((resolve) => {
    resolvePromise = resolve;
  });
  const callback = (e) => {
    const isValidOrigin = supportedOrigins.some((o) => e.origin === o);
    if (!isValidOrigin) return;
    const data = e.data;
    if (!data || typeof data !== "object") return;
    if (data.type !== EXPECTED_MESSAGE_TYPE) return;
    resolvePromise(data.response);
  };
  window.addEventListener("message", callback);
  return {
    cleanup: () => window.removeEventListener("message", callback),
    messagePromise: promise
  };
}
function processOAuthResponse(data, expectedState) {
  if (data.state !== expectedState) {
    return { error: new Error("State is invalid") };
  }
  if (data.error) {
    return {
      error: new Error(data.error_description ?? "Sign in failed")
    };
  }
  if (!data.access_token || !data.refresh_token) {
    return { error: new Error("No tokens received") };
  }
  return {
    tokens: {
      access_token: data.access_token,
      refresh_token: data.refresh_token
    },
    error: null
  };
}
function createGretaAuth(config = {}) {
  const oauthBrokerUrl = config.oauthBrokerUrl ?? DEFAULT_OAUTH_BROKER_URL;
  const supportedOAuthOrigins = config.supportedOAuthOrigins ?? DEFAULT_SUPPORTED_OAUTH_ORIGINS;
  async function signInWithOAuth(provider, opts = {}) {
    let isInIframe = false;
    try {
      isInIframe = window.self !== window.top;
    } catch {
      isInIframe = true;
    }
    const ua = navigator.userAgent;
    const isMobileApp = /GretaApp\//.test(ua);
    const isDesktopApp = !isMobileApp && /greta/i.test(ua);
    const state = generateState();
    let redirectUri = opts.redirect_uri ?? window.location.origin;
    if (isMobileApp && isInIframe) {
      redirectUri = DEFAULT_MOBILE_DEEP_LINK_REDIRECT_URI;
    } else if (isDesktopApp && isInIframe) {
      redirectUri = DEFAULT_DESKTOP_LOCALHOST_REDIRECT_URI;
    }
    const params = new URLSearchParams({
      ...opts.extraParams,
      provider,
      redirect_uri: redirectUri,
      state
    });
    if (!isInIframe) {
      window.location.href = `${oauthBrokerUrl}?${params.toString()}`;
      return { error: null, redirected: true };
    }
    if (!isMobileApp && !isDesktopApp) {
      params.set("response_mode", "web_message");
    }
    const url = `${oauthBrokerUrl}?${params.toString()}`;
    const effectiveOrigins = !isMobileApp && !isDesktopApp ? [window.location.origin, ...supportedOAuthOrigins] : isDesktopApp ? [...supportedOAuthOrigins, window.location.origin] : supportedOAuthOrigins;
    const { messagePromise, cleanup } = startWebMessageListener(effectiveOrigins);
    let popup;
    if (isDevice()) {
      popup = window.open(url, "_blank");
    } else {
      const { width, height, left, top } = getPopupDimensions(isInIframe);
      popup = window.open(
        url,
        "oauth",
        `width=${width},height=${height},left=${left},top=${top}`
      );
    }
    if (!popup && (isMobileApp || isDesktopApp)) {
      let webViewTimeoutId;
      const webViewTimeoutPromise = new Promise((_, reject) => {
        webViewTimeoutId = setTimeout(
          () => reject(new Error("OAuth timed out waiting for response")),
          IFRAME_FALLBACK_TIMEOUT_MS
        );
      });
      try {
        const data = await Promise.race([messagePromise, webViewTimeoutPromise]);
        return processOAuthResponse(data, state);
      } catch (error) {
        return { error: error instanceof Error ? error : new Error(String(error)) };
      } finally {
        clearTimeout(webViewTimeoutId);
        cleanup();
      }
    }
    if (!popup) {
      cleanup();
      return { error: new Error("Popup was blocked") };
    }
    let popupCheckInterval;
    const popupClosedPromise = new Promise((_, reject) => {
      popupCheckInterval = setInterval(() => {
        if (popup.closed) {
          clearInterval(popupCheckInterval);
          reject(new Error("Sign in was cancelled"));
        }
      }, POPUP_CHECK_INTERVAL_MS);
    });
    try {
      const data = await Promise.race([messagePromise, popupClosedPromise]);
      return processOAuthResponse(data, state);
    } catch (error) {
      return { error: error instanceof Error ? error : new Error(String(error)) };
    } finally {
      clearInterval(popupCheckInterval);
      cleanup();
      popup?.close();
    }
  }
  return { signInWithOAuth };
}

// src/react/context.tsx
import {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useState
} from "react";
import { jsx } from "react/jsx-runtime";
var SESSION_KEY = "google_auth_session";
var GoogleAuthContext = createContext(null);
async function fetchGoogleUser(accessToken) {
  const res = await fetch("https://www.googleapis.com/oauth2/v3/userinfo", {
    headers: { Authorization: `Bearer ${accessToken}` }
  });
  if (!res.ok) throw new Error("Failed to fetch Google user info");
  return res.json();
}
function GoogleAuthProvider({
  children,
  config
}) {
  const [user, setUser] = useState(null);
  const [tokens, setTokens] = useState(null);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(null);
  const [initialized, setInitialized] = useState(false);
  useEffect(() => {
    async function init() {
      const hash = new URLSearchParams(window.location.hash.slice(1));
      const accessToken = hash.get("access_token");
      const refreshToken = hash.get("refresh_token");
      if (accessToken) {
        window.history.replaceState(null, "", window.location.pathname + window.location.search);
        try {
          setLoading(true);
          const googleUser = await fetchGoogleUser(accessToken);
          const t = { access_token: accessToken, refresh_token: refreshToken ?? "" };
          const session = { tokens: t, user: googleUser };
          sessionStorage.setItem(SESSION_KEY, JSON.stringify(session));
          setTokens(t);
          setUser(googleUser);
        } catch {
        } finally {
          setLoading(false);
          setInitialized(true);
        }
        return;
      }
      try {
        const raw = sessionStorage.getItem(SESSION_KEY);
        if (raw) {
          const { tokens: t, user: u } = JSON.parse(raw);
          setTokens(t);
          setUser(u);
        }
      } catch {
        sessionStorage.removeItem(SESSION_KEY);
      } finally {
        setInitialized(true);
      }
    }
    init();
  }, []);
  const signIn = useCallback(
    async (options) => {
      setLoading(true);
      setError(null);
      try {
        const auth = createGretaAuth(config);
        const result = await auth.signInWithOAuth("google", options);
        if (result.redirected) return;
        if (result.error) {
          setError(result.error);
          return;
        }
        console.log("\u2705 Google tokens received:");
        console.log("  access_token :", result.tokens.access_token);
        console.log("  refresh_token:", result.tokens.refresh_token);
        const googleUser = await fetchGoogleUser(result.tokens.access_token);
        const session = {
          tokens: result.tokens,
          user: googleUser
        };
        sessionStorage.setItem(SESSION_KEY, JSON.stringify(session));
        setTokens(result.tokens);
        setUser(googleUser);
      } catch (err) {
        setError(err instanceof Error ? err : new Error(String(err)));
      } finally {
        setLoading(false);
      }
    },
    [config]
  );
  const signOut = useCallback(() => {
    sessionStorage.removeItem(SESSION_KEY);
    setUser(null);
    setTokens(null);
    setError(null);
  }, []);
  const value = {
    user,
    tokens,
    loading,
    error,
    initialized,
    signIn,
    signOut
  };
  return /* @__PURE__ */ jsx(GoogleAuthContext.Provider, { value, children });
}
function useGoogleAuth() {
  const ctx = useContext(GoogleAuthContext);
  if (!ctx) {
    throw new Error("useGoogleAuth must be used within a GoogleAuthProvider");
  }
  return ctx;
}

// src/react/useGoogleSignIn.ts
import { useCallback as useCallback2, useState as useState2 } from "react";
function useGoogleSignIn(hookOptions) {
  const [status, setStatus] = useState2("idle");
  const [error, setError] = useState2(null);
  const signIn = useCallback2(
    async (options) => {
      setStatus("loading");
      setError(null);
      try {
        const result = await createGretaAuth(hookOptions?.config).signInWithOAuth("google", options);
        if (result.redirected) {
          return;
        }
        if (result.error) {
          setStatus("error");
          setError(result.error);
          hookOptions?.onError?.(result.error);
          return;
        }
        setStatus("success");
        hookOptions?.onSuccess?.(result.tokens);
      } catch (err) {
        const e = err instanceof Error ? err : new Error(String(err));
        setStatus("error");
        setError(e);
        hookOptions?.onError?.(e);
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [hookOptions?.config, hookOptions?.onSuccess, hookOptions?.onError]
  );
  const reset = useCallback2(() => {
    setStatus("idle");
    setError(null);
  }, []);
  return {
    signIn,
    status,
    loading: status === "loading",
    error,
    reset
  };
}

// src/react/GoogleSignInButton.tsx
import { jsx as jsx2, jsxs } from "react/jsx-runtime";
function GoogleIcon({ size = 18 }) {
  return /* @__PURE__ */ jsxs(
    "svg",
    {
      xmlns: "http://www.w3.org/2000/svg",
      viewBox: "0 0 48 48",
      width: size,
      height: size,
      "aria-hidden": "true",
      children: [
        /* @__PURE__ */ jsx2(
          "path",
          {
            fill: "#EA4335",
            d: "M24 9.5c3.54 0 6.71 1.22 9.21 3.6l6.85-6.85C35.9 2.38 30.47 0 24 0 14.62 0 6.51 5.38 2.56 13.22l7.98 6.19C12.43 13.72 17.74 9.5 24 9.5z"
          }
        ),
        /* @__PURE__ */ jsx2(
          "path",
          {
            fill: "#4285F4",
            d: "M46.98 24.55c0-1.57-.15-3.09-.38-4.55H24v9.02h12.94c-.58 2.96-2.26 5.48-4.78 7.18l7.73 6c4.51-4.18 7.09-10.36 7.09-17.65z"
          }
        ),
        /* @__PURE__ */ jsx2(
          "path",
          {
            fill: "#FBBC05",
            d: "M10.53 28.59c-.48-1.45-.76-2.99-.76-4.59s.27-3.14.76-4.59l-7.98-6.19C.92 16.46 0 20.12 0 24c0 3.88.92 7.54 2.56 10.78l7.97-6.19z"
          }
        ),
        /* @__PURE__ */ jsx2(
          "path",
          {
            fill: "#34A853",
            d: "M24 48c6.48 0 11.93-2.13 15.89-5.81l-7.73-6c-2.18 1.48-4.97 2.36-8.16 2.36-6.26 0-11.57-4.22-13.47-9.91l-7.98 6.19C6.51 42.62 14.62 48 24 48z"
          }
        )
      ]
    }
  );
}
var BASE = {
  display: "inline-flex",
  alignItems: "center",
  justifyContent: "center",
  border: "none",
  padding: 0,
  cursor: "pointer",
  fontFamily: "'Roboto', Arial, sans-serif",
  fontSize: "14px",
  fontWeight: 500,
  letterSpacing: "0.25px",
  lineHeight: 1,
  userSelect: "none",
  whiteSpace: "nowrap",
  outline: "none",
  transition: "background-color 0.218s, box-shadow 0.218s",
  height: "40px"
};
var LIGHT = {
  backgroundColor: "#FFFFFF",
  color: "#3C4043",
  border: "1px solid #DADCE0",
  boxShadow: "0 1px 2px rgba(60,64,67,0.3), 0 1px 3px 1px rgba(60,64,67,0.15)"
};
var DARK = {
  backgroundColor: "#131314",
  color: "#E3E3E3",
  border: "1px solid #8E918F"
};
var ICON_WRAPPER_DARK = {
  display: "flex",
  alignItems: "center",
  justifyContent: "center",
  width: "20px",
  height: "20px",
  backgroundColor: "#FFFFFF",
  borderRadius: "2px"
};
function GoogleSignInButton({
  theme = "light",
  shape = "rectangular",
  variant = "standard",
  text = "Sign in with Google",
  onSuccess,
  onError,
  config,
  options,
  disabled,
  className,
  style
}) {
  const { signIn, loading } = useGoogleSignIn({
    config,
    onSuccess,
    onError
  });
  const isIcon = variant === "icon";
  const isLight = theme === "light";
  const borderRadius = shape === "pill" ? "20px" : "4px";
  const buttonStyle = {
    ...BASE,
    ...isLight ? LIGHT : DARK,
    borderRadius,
    width: isIcon ? "40px" : void 0,
    padding: isIcon ? "0" : "0 12px 0 0",
    opacity: disabled || loading ? 0.6 : 1,
    cursor: disabled || loading ? "not-allowed" : "pointer",
    ...style
  };
  const logoContainerStyle = {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    width: "40px",
    height: "40px",
    flexShrink: 0
  };
  const handleClick = () => {
    if (!disabled && !loading) signIn(options);
  };
  const label = loading ? "Signing in\u2026" : text;
  if (isIcon) {
    return /* @__PURE__ */ jsx2(
      "button",
      {
        type: "button",
        style: buttonStyle,
        className,
        disabled: disabled || loading,
        onClick: handleClick,
        "aria-label": label,
        children: isLight ? /* @__PURE__ */ jsx2(GoogleIcon, { size: 18 }) : /* @__PURE__ */ jsx2("span", { style: ICON_WRAPPER_DARK, children: /* @__PURE__ */ jsx2(GoogleIcon, { size: 18 }) })
      }
    );
  }
  return /* @__PURE__ */ jsxs(
    "button",
    {
      type: "button",
      style: buttonStyle,
      className,
      disabled: disabled || loading,
      onClick: handleClick,
      "aria-label": label,
      children: [
        /* @__PURE__ */ jsx2("span", { style: logoContainerStyle, children: isLight ? /* @__PURE__ */ jsx2(GoogleIcon, { size: 18 }) : /* @__PURE__ */ jsx2("span", { style: ICON_WRAPPER_DARK, children: /* @__PURE__ */ jsx2(GoogleIcon, { size: 18 }) }) }),
        /* @__PURE__ */ jsx2("span", { style: { paddingRight: "12px" }, children: label })
      ]
    }
  );
}

// src/react/RequireAuth.tsx
import { Fragment, jsx as jsx3 } from "react/jsx-runtime";
function RequireAuth({
  children,
  fallback = null,
  unauthenticated = null
}) {
  const { initialized, loading, user } = useGoogleAuth();
  if (!initialized || loading) {
    return /* @__PURE__ */ jsx3(Fragment, { children: fallback });
  }
  if (!user) {
    return /* @__PURE__ */ jsx3(Fragment, { children: unauthenticated });
  }
  return /* @__PURE__ */ jsx3(Fragment, { children });
}

// src/react/withGoogleAuth.tsx
import { jsx as jsx4 } from "react/jsx-runtime";
function withGoogleAuth(Component) {
  const displayName = Component.displayName ?? Component.name ?? "Component";
  function WithGoogleAuthWrapper(props) {
    const auth = useGoogleAuth();
    return /* @__PURE__ */ jsx4(Component, { ...props, googleAuth: auth });
  }
  WithGoogleAuthWrapper.displayName = `withGoogleAuth(${displayName})`;
  return WithGoogleAuthWrapper;
}

// src/react/useGretaAuthCallback.ts
import { useEffect as useEffect2 } from "react";
function useGretaAuthCallback(onUserOrHandlers) {
  useEffect2(() => {
    const handlers = typeof onUserOrHandlers === "function" ? { onUser: onUserOrHandlers } : onUserOrHandlers;
    const hash = new URLSearchParams(window.location.hash.slice(1));
    const accessToken = hash.get("access_token");
    if (!accessToken) {
      handlers.onDone?.();
      return;
    }
    window.history.replaceState(null, "", window.location.pathname);
    fetch("https://www.googleapis.com/oauth2/v3/userinfo", {
      headers: { Authorization: `Bearer ${accessToken}` }
    }).then((r) => r.json()).then((user) => {
      handlers.onUser(user, accessToken);
      handlers.onDone?.();
    }).catch((e) => {
      console.error(e);
      handlers.onDone?.();
    });
  }, []);
}
export {
  DEFAULT_DESKTOP_LOCALHOST_REDIRECT_URI,
  DEFAULT_MOBILE_DEEP_LINK_REDIRECT_URI,
  DEFAULT_OAUTH_BROKER_URL,
  DEFAULT_SUPPORTED_OAUTH_ORIGINS,
  EXPECTED_MESSAGE_TYPE,
  GoogleAuthProvider,
  GoogleSignInButton,
  IFRAME_FALLBACK_TIMEOUT_MS,
  POPUP_CHECK_INTERVAL_MS,
  RequireAuth,
  createGretaAuth,
  generateState,
  getPopupDimensions,
  isDevice,
  processOAuthResponse,
  startWebMessageListener,
  useGoogleAuth,
  useGoogleSignIn,
  useGretaAuthCallback,
  withGoogleAuth
};
