// @questlabs/auth-js/dist/broker/server.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// broker/server.ts
var server_exports = {};
__export(server_exports, {
  default: () => server_default
});
module.exports = __toCommonJS(server_exports);
var import_crypto = __toESM(require("crypto"));
var import_express = __toESM(require("express"));
var app = (0, import_express.default)();
var GOOGLE_AUTH_URL = "https://accounts.google.com/o/oauth2/v2/auth";
var GOOGLE_TOKEN_URL = "https://oauth2.googleapis.com/token";
var SESSION_TTL_MS = 5 * 60 * 1e3;
var sessions = /* @__PURE__ */ new Map();
setInterval(() => {
  const cutoff = Date.now() - SESSION_TTL_MS;
  for (const [id, s] of sessions) {
    if (s.createdAt < cutoff) sessions.delete(id);
  }
}, 6e4).unref();
function generateCodeVerifier() {
  return import_crypto.default.randomBytes(96).toString("base64url");
}
function generateCodeChallenge(verifier) {
  return import_crypto.default.createHash("sha256").update(verifier).digest("base64url");
}
function generateSessionId() {
  return import_crypto.default.randomBytes(16).toString("hex");
}
app.get("/~oauth/initiate", (req, res) => {
  const {
    provider,
    state,
    redirect_uri,
    client_id,
    client_secret,
    mode = "web_message",
    ...extraParams
  } = req.query;
  if (provider !== "google") {
    res.status(400).json({ error: `Unsupported provider: ${provider}` });
    return;
  }
  const clientId = client_id || process.env.GOOGLE_CLIENT_ID;
  const clientSecret = client_secret || process.env.GOOGLE_CLIENT_SECRET;
  if (!clientId || !clientSecret) {
    res.status(400).json({
      error: "Missing Google credentials. Set GOOGLE_CLIENT_ID / GOOGLE_CLIENT_SECRET or pass client_id + client_secret as query params (BYOK mode)."
    });
    return;
  }
  const codeVerifier = generateCodeVerifier();
  const codeChallenge = generateCodeChallenge(codeVerifier);
  const sessionId = generateSessionId();
  const baseUrl = `${req.protocol}://${req.get("host")}`;
  const callbackRedirectUri = `${baseUrl}/~oauth/callback`;
  sessions.set(sessionId, {
    codeVerifier,
    clientState: state || "",
    clientId,
    clientSecret,
    callbackRedirectUri,
    deliveryMode: mode,
    originalRedirectUri: redirect_uri || "/",
    createdAt: Date.now()
  });
  const googleParams = new URLSearchParams({
    client_id: clientId,
    redirect_uri: callbackRedirectUri,
    response_type: "code",
    scope: "openid email profile",
    state: sessionId,
    code_challenge: codeChallenge,
    code_challenge_method: "S256",
    access_type: "offline",
    prompt: "consent",
    ...extraParams
  });
  res.redirect(`${GOOGLE_AUTH_URL}?${googleParams.toString()}`);
});
app.get("/~oauth/callback", async (req, res) => {
  const { code, state: sessionId, error } = req.query;
  if (error) {
    res.status(400).send(errorPage(error));
    return;
  }
  const session = sessions.get(sessionId);
  if (!session) {
    res.status(400).send(errorPage("Invalid or expired OAuth session."));
    return;
  }
  sessions.delete(sessionId);
  try {
    const tokenRes = await fetch(GOOGLE_TOKEN_URL, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams({
        code,
        client_id: session.clientId,
        client_secret: session.clientSecret,
        redirect_uri: session.callbackRedirectUri,
        grant_type: "authorization_code",
        code_verifier: session.codeVerifier
      }).toString()
    });
    const tokenData = await tokenRes.json();
    if (!tokenRes.ok || !tokenData.access_token) {
      throw new Error(
        tokenData.error_description ?? tokenData.error ?? "Token exchange failed"
      );
    }
    if (session.deliveryMode === "web_message") {
      res.send(webMessagePage(session.clientState, tokenData.access_token, tokenData.refresh_token ?? ""));
      return;
    }
    const hash = new URLSearchParams({
      access_token: tokenData.access_token,
      refresh_token: tokenData.refresh_token ?? "",
      state: session.clientState
    }).toString();
    res.redirect(`${session.originalRedirectUri}#${hash}`);
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    res.status(500).send(errorPage(message));
  }
});
function webMessagePage(state, access_token, refresh_token) {
  const payload = JSON.stringify({
    type: "authorization_response",
    response: { access_token, refresh_token, state }
  });
  return `<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8"><title>Authenticating\u2026</title></head>
<body>
<script>
(function () {
  var payload = ${payload};
  if (window.opener) {
    window.opener.postMessage(payload, '*');
    window.close();
  } else if (window.parent !== window) {
    window.parent.postMessage(payload, '*');
  }
})();
</script>
<p style="font-family:sans-serif">Authentication complete. You may close this window.</p>
</body>
</html>`;
}
function errorPage(message) {
  const safe = message.replace(/</g, "&lt;").replace(/>/g, "&gt;");
  return `<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8"><title>OAuth Error</title></head>
<body>
<script>
(function () {
  var payload = {
    type: 'authorization_response',
    response: { error: ${JSON.stringify(safe)}, state: '' }
  };
  if (window.opener) { window.opener.postMessage(payload, '*'); window.close(); }
})();
</script>
<p style="font-family:sans-serif;color:#c00">OAuth error: ${safe}</p>
</body>
</html>`;
}
var PORT = Number(process.env.PORT ?? 3001);
app.listen(PORT, () => {
  console.log(`OAuth broker listening on http://localhost:${PORT}`);
});
var server_default = app;
